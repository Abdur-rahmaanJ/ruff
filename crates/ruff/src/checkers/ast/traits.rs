use ruff_diagnostics::Diagnostic;
use rustpython_parser::ast;

use crate::checkers::ast::ImmutableChecker;

pub(crate) struct RegistryRule {
    pub(crate) run: RuleExecutor,
}

type RuleExecutor =
    fn(diagnostics: &mut Vec<Diagnostic>, checker: &ImmutableChecker, node: &ast::ExprCall);

impl RegistryRule {
    pub(crate) fn new<R: AnalysisRule + 'static>() -> Self {
        Self { run: R::run }
    }

    pub fn run(
        &self,
        diagnostics: &mut Vec<Diagnostic>,
        checker: &ImmutableChecker,
        node: &ast::ExprCall,
    ) {
        (self.run)(diagnostics, checker, node)
    }
}

pub(crate) trait AnalysisRule: Sized {
    fn run(diagnostics: &mut Vec<Diagnostic>, checker: &ImmutableChecker, node: &ast::ExprCall);
}
