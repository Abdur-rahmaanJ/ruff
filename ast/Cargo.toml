[package]
name = "rustpython-ast"
version = "0.2.0"
description = "AST definitions for RustPython"
authors = ["RustPython Team"]
edition = "2021"
repository = "https://github.com/RustPython/RustPython"
license = "MIT"

[features]
default = ["constant-optimization", "fold", "source-code"]
constant-optimization = ["fold"]
source-code = ["fold"]
fold = []
unparse = ["rustpython-literal"]

[dependencies]
rustpython-parser-core = { workspace = true }
rustpython-literal = { workspace = true, optional = true }

num-bigint = { workspace = true }
