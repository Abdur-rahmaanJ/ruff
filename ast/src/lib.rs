mod attributed;
mod constant;
#[cfg(feature = "fold")]
mod fold_helpers;
mod generic {
    #![allow(clippy::derive_partial_eq_without_eq)]
    include!("gen/generic.rs");
}
mod impls;
#[cfg(feature = "location")]
pub mod located {
    include!("gen/located.rs");
}
#[cfg(feature = "location")]
mod locator;
#[cfg(feature = "location")]
pub use crate::locator::locate;
#[cfg(feature = "location")]
pub use rustpython_compiler_core::SourceLocator;

#[cfg(feature = "unparse")]
mod unparse;

pub use attributed::Attributed;
pub use constant::{Constant, ConversionFlag};
pub use generic::*;

pub type Suite<U = ()> = Vec<Stmt<U>>;

pub mod location {
    pub use rustpython_compiler_core::source_code::{OneIndexed, SourceLocation};

    #[derive(Debug)]
    pub struct SourceRange {
        pub start: SourceLocation,
        pub end: Option<SourceLocation>,
    }

    impl SourceRange {
        pub fn new(start: SourceLocation, end: SourceLocation) -> Self {
            Self {
                start,
                end: Some(end),
            }
        }
        pub fn unwrap_end(&self) -> SourceLocation {
            self.end.unwrap()
        }
    }

    impl From<std::ops::Range<SourceLocation>> for SourceRange {
        fn from(value: std::ops::Range<SourceLocation>) -> Self {
            Self {
                start: value.start,
                end: Some(value.end),
            }
        }
    }
}
