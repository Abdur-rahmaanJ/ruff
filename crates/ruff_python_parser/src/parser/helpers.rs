use ruff_python_ast::{self as ast, Expr, ExprContext};

/// Set the `ctx` for `Expr::Id`, `Expr::Attribute`, `Expr::Subscript`, `Expr::Starred`,
/// `Expr::Tuple` and `Expr::List`. If `expr` is either `Expr::Tuple` or `Expr::List`,
/// recursively sets the `ctx` for their elements.
pub(super) fn set_expr_ctx(expr: &mut Expr, new_ctx: ExprContext) {
    match expr {
        Expr::Name(ast::ExprName { ctx, .. })
        | Expr::Attribute(ast::ExprAttribute { ctx, .. })
        | Expr::Subscript(ast::ExprSubscript { ctx, .. }) => *ctx = new_ctx,
        Expr::Starred(ast::ExprStarred { value, ctx, .. }) => {
            *ctx = new_ctx;
            set_expr_ctx(value, new_ctx);
        }
        Expr::UnaryOp(ast::ExprUnaryOp { operand, .. }) => {
            set_expr_ctx(operand, new_ctx);
        }
        Expr::List(ast::ExprList { elts, ctx, .. })
        | Expr::Tuple(ast::ExprTuple { elts, ctx, .. }) => {
            *ctx = new_ctx;
            elts.iter_mut()
                .for_each(|element| set_expr_ctx(element, new_ctx));
        }
        _ => {}
    }
}
