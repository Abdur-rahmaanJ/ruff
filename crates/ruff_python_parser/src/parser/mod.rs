use bitflags::bitflags;

use ast::Mod;
use ruff_python_ast::{self as ast, Expr, Stmt};
use ruff_text_size::{Ranged, TextLen, TextRange, TextSize};

use crate::lexer::lex;
use crate::parser::progress::ParserProgress;
use crate::{
    lexer::{LexResult, Spanned},
    token_set::TokenSet,
    token_source::TokenSource,
    Mode, ParseError, ParseErrorType, Tok, TokenKind,
};

mod expression;
mod helpers;
mod pattern;
mod progress;
mod statement;
#[cfg(test)]
mod tests;

pub(crate) fn parse_tokens(
    tokens: Vec<LexResult>,
    source: &str,
    mode: Mode,
) -> Result<Mod, ParseError> {
    let program = Parser::new(source, mode, TokenSource::new(tokens)).parse_program();
    if program.parse_errors.is_empty() {
        Ok(program.ast)
    } else {
        Err(program.parse_errors.into_iter().next().unwrap())
    }
}

#[derive(Debug)]
pub struct Program {
    pub ast: ast::Mod,
    pub parse_errors: Vec<ParseError>,
}

impl Program {
    pub fn parse_str(source: &str, mode: Mode) -> Program {
        let tokens = lex(source, mode);
        Self::parse_tokens(source, tokens.collect(), mode)
    }

    pub fn parse_tokens(source: &str, tokens: Vec<LexResult>, mode: Mode) -> Program {
        Parser::new(source, mode, TokenSource::new(tokens)).parse_program()
    }
}

pub(crate) struct Parser<'src> {
    source: &'src str,
    tokens: TokenSource,

    /// Stores all the syntax errors found during the parsing.
    errors: Vec<ParseError>,

    /// This tracks the current expression or statement being parsed. For example,
    /// if we're parsing a tuple expression, e.g. `(1, 2)`, `ctx` has the value
    /// `ParserCtxFlags::TUPLE_EXPR`.
    ///
    /// The `ctx` is also used to create custom error messages and forbid certain
    /// expressions or statements of being parsed. The `ctx` should be empty after
    /// an expression or statement is done parsing.
    ctx: ParserCtxFlags,

    /// During the parsing of expression or statement, multiple `ctx`s can be created.
    /// `ctx_stack` stores the previous `ctx`s that were created during the parsing. For example,
    /// when parsing a tuple expression, e.g. `(1, 2, 3)`, two [`ParserCtxFlags`] will be
    /// created `ParserCtxFlags::PARENTHESIZED_EXPR` and `ParserCtxFlags::TUPLE_EXPR`.
    ///
    /// When parsing a tuple the first context created is `ParserCtxFlags::PARENTHESIZED_EXPR`.
    /// Afterwards, the `ParserCtxFlags::TUPLE_EXPR` is created and `ParserCtxFlags::PARENTHESIZED_EXPR`
    /// is pushed onto the `ctx_stack`.
    /// `ParserCtxFlags::PARENTHESIZED_EXPR` is removed from the stack and set to be the current `ctx`,
    /// after we parsed all elements in the tuple.
    ///
    /// The end of the vector is the top of the stack.
    ctx_stack: Vec<ParserCtxFlags>,

    /// Stores the last `ctx` of an expression or statement that was parsed.
    last_ctx: ParserCtxFlags,

    /// Specify the mode in which the code will be parsed.
    mode: Mode,

    /// Defer the creation of the invalid node for the skipped unexpected tokens.
    /// Holds the range of the skipped tokens.
    defer_invalid_node_creation: Option<TextRange>,

    current: Spanned,

    /// The end of the last processed. Used to determine a node's end.
    last_token_end: TextSize,
}

// TODO: Review use of `ParsedExpr`: Can we reduce usage?

const NEWLINE_EOF_SET: TokenSet = TokenSet::new(&[TokenKind::Newline, TokenKind::EndOfFile]);
const LITERAL_SET: TokenSet = TokenSet::new(&[
    TokenKind::Name,
    TokenKind::Int,
    TokenKind::Float,
    TokenKind::Complex,
    TokenKind::Plus,
    TokenKind::String,
    TokenKind::Ellipsis,
    TokenKind::True,
    TokenKind::False,
    TokenKind::None,
]);
/// Tokens that are usually an expression or the start of one.
const EXPR_SET: TokenSet = TokenSet::new(&[
    TokenKind::Minus,
    TokenKind::Tilde,
    TokenKind::Star,
    TokenKind::DoubleStar,
    TokenKind::Vbar,
    TokenKind::Lpar,
    TokenKind::Lbrace,
    TokenKind::Lsqb,
    TokenKind::Lambda,
    TokenKind::Await,
    TokenKind::Not,
    TokenKind::Yield,
    TokenKind::FStringStart,
])
.union(LITERAL_SET);

impl<'src> Parser<'src> {
    pub(crate) fn new(source: &'src str, mode: Mode, mut tokens: TokenSource) -> Parser<'src> {
        let current = tokens
            .next()
            .unwrap_or_else(|| (Tok::EndOfFile, TextRange::empty(source.text_len())));

        Parser {
            mode,
            source,
            errors: Vec::new(),
            ctx_stack: Vec::new(),
            ctx: ParserCtxFlags::empty(),
            last_ctx: ParserCtxFlags::empty(),
            tokens,
            last_token_end: TextSize::default(),
            current,

            defer_invalid_node_creation: None,
        }
    }

    pub(crate) fn parse_program(mut self) -> Program {
        let mut body = vec![];

        let ast = if self.mode == Mode::Expression {
            let start = self.node_start();
            let parsed_expr = self.parse_expression();
            loop {
                if !self.eat(TokenKind::Newline) {
                    break;
                }
            }
            self.expect(TokenKind::EndOfFile);

            ast::Mod::Expression(ast::ModExpression {
                body: Box::new(parsed_expr.expr),
                range: self.node_range(start),
            })
        } else {
            let is_src_empty = self.at(TokenKind::EndOfFile);
            while !self.at(TokenKind::EndOfFile) {
                if self.at(TokenKind::Indent) {
                    self.handle_unexpected_indentation(&mut body, "unexpected indentation");
                    continue;
                }

                body.push(self.parse_statement());

                if let Some(range) = self.defer_invalid_node_creation {
                    self.defer_invalid_node_creation = None;
                    body.push(Stmt::Expr(ast::StmtExpr {
                        value: Box::new(Expr::Invalid(ast::ExprInvalid {
                            value: self.src_text(range).into(),
                            range,
                        })),
                        range,
                    }));
                }
            }
            ast::Mod::Module(ast::ModModule {
                body,
                // If the `source` only contains comments or empty spaces, return
                // an empty range.
                // FIXME(micha): The modul should always enclose the entire file
                range: if is_src_empty {
                    TextRange::default()
                } else {
                    TextRange::new(
                        0.into(),
                        self.source
                            .len()
                            .try_into()
                            .expect("source length is  bigger than u32 max"),
                    )
                },
            })
        };

        Program {
            ast,
            parse_errors: self.finish(),
        }
    }

    fn finish(self) -> Vec<ParseError> {
        // After parsing, the `ctx` and `ctx_stack` should be empty.
        // If it's not, you probably forgot to call `clear_ctx` somewhere.
        assert_eq!(self.ctx, ParserCtxFlags::empty());
        assert_eq!(&self.ctx_stack, &[]);
        assert_eq!(
            self.current,
            (Tok::EndOfFile, TextRange::empty(self.source.text_len())),
            "Parser should be at the end of the file."
        );

        // TODO consider re-integrating lexical error handling into the parser?
        let parse_errors = self.errors;
        let lex_errors = self.tokens.finish();

        // Fast path for when there are no lex errors.
        // There's no fast path for when there are no parse errors because a lex error
        // always results in a parse error.
        if lex_errors.is_empty() {
            return parse_errors;
        }

        let mut merged = Vec::with_capacity(parse_errors.len().saturating_add(lex_errors.len()));

        let mut parse_errors = parse_errors.into_iter().peekable();
        let mut lex_errors = lex_errors.into_iter().peekable();

        while let (Some(parse_error), Some(lex_error)) = (parse_errors.peek(), lex_errors.peek()) {
            if parse_error.location.start() < lex_error.location.start() {
                merged.push(parse_errors.next().unwrap());
            } else {
                merged.push(ParseError::from(lex_errors.next().unwrap()));
            }
        }

        merged.extend(parse_errors);
        merged.extend(lex_errors.map(ParseError::from));

        merged
    }

    #[inline]
    fn set_ctx(&mut self, ctx: ParserCtxFlags) {
        self.ctx_stack.push(self.ctx);
        self.ctx = ctx;
    }

    #[inline]
    fn clear_ctx(&mut self, ctx: ParserCtxFlags) {
        assert_eq!(self.ctx, ctx);
        self.last_ctx = ctx;
        if let Some(top) = self.ctx_stack.pop() {
            self.ctx = top;
        }
    }

    #[inline]
    fn has_ctx(&self, ctx: ParserCtxFlags) -> bool {
        self.ctx.intersects(ctx)
    }

    /// Returns the start position for a node that starts at the current token.
    fn node_start(&self) -> TextSize {
        self.current_range().start()
    }

    fn node_range(&self, start: TextSize) -> TextRange {
        TextRange::new(start, self.last_token_end)
    }

    /// Moves the parser to the next token. Returns the old current token as an owned value.
    fn next_token(&mut self) -> Spanned {
        let next = self
            .tokens
            .next()
            .unwrap_or_else(|| (Tok::EndOfFile, TextRange::empty(self.source.text_len())));

        let current = std::mem::replace(&mut self.current, next);

        if !matches!(
            current.0,
            // TODO explore including everything up to the dedent as part of the body.
            Tok::Dedent
            // Don't include newlines in the body
            | Tok::Newline
            // TODO(micha): Including the semi feels more correct but it isn't compatible with lalrpop and breaks the
            // formatters semicolon detection. Exclude it for now
            | Tok::Semi
        ) {
            self.last_token_end = current.1.end();
        }

        current
    }

    fn peek_nth(&mut self, offset: usize) -> TokenKind {
        if offset == 0 {
            self.current_kind()
        } else {
            self.tokens
                .peek_nth(offset - 1)
                .map_or(TokenKind::EndOfFile, |spanned| spanned.0)
        }
    }

    #[inline]
    fn current_token(&self) -> (TokenKind, TextRange) {
        (self.current_kind(), self.current_range())
    }

    #[inline]
    fn current_kind(&self) -> TokenKind {
        // TODO: Converting the token kind over and over again can be expensive.
        TokenKind::from_token(&self.current.0)
    }

    #[inline]
    fn current_range(&self) -> TextRange {
        self.current.1
    }

    fn eat(&mut self, kind: TokenKind) -> bool {
        if !self.at(kind) {
            return false;
        }

        self.next_token();
        true
    }

    /// Bumps the current token assuming it is of the given kind.
    ///
    /// # Panics
    /// If the current token is not of the given kind.
    ///
    /// # Returns
    /// The current token
    fn bump(&mut self, kind: TokenKind) -> Spanned {
        assert_eq!(self.current_kind(), kind);

        self.next_token()
    }

    fn expect(&mut self, expected: TokenKind) -> bool {
        if self.eat(expected) {
            return true;
        }

        let (found, range) = self.current_token();
        self.add_error(ParseErrorType::ExpectedToken { found, expected }, range);
        false
    }

    /// Expects a specific token kind, skipping leading unexpected tokens if needed.
    fn expect_and_recover(&mut self, expected: TokenKind, recover_set: TokenSet) {
        if !self.expect(expected) {
            let expected_set = NEWLINE_EOF_SET
                .union(recover_set)
                .union([expected].as_slice().into());
            // Skip leading unexpected tokens
            let range = self.skip_until(expected_set);
            self.defer_invalid_node_creation = Some(range);

            self.add_error(
                ParseErrorType::OtherError("unexpected tokens".into()),
                range,
            );

            self.eat(expected);
        }
    }

    fn add_error<T>(&mut self, error: ParseErrorType, ranged: T)
    where
        T: Ranged,
    {
        self.errors.push(ParseError {
            error,
            location: ranged.range(),
        });
    }

    /// Skip tokens until [`TokenSet`]. Returns the range of the skipped tokens.
    fn skip_until(&mut self, token_set: TokenSet) -> TextRange {
        let mut final_range = self.current_range();
        while !self.at_ts(token_set) {
            let (_, range) = self.next_token();
            final_range = final_range.cover(range);
        }

        final_range
    }

    fn at(&mut self, kind: TokenKind) -> bool {
        self.current_kind() == kind
    }

    fn at_ts(&mut self, ts: TokenSet) -> bool {
        ts.contains(self.current_kind())
    }

    fn src_text<T>(&self, ranged: T) -> &'src str
    where
        T: Ranged,
    {
        // This check is to prevent the parser from panicking when using the
        // `parse_expression_starts_at` function with an offset bigger than zero.
        //
        // The parser assumes that the token's range values are smaller than
        // the source length. But, with an offset bigger than zero, it can
        // happen that the token's range values are bigger than the source
        // length, causing the parser to panic when calling this function
        // with such ranges.
        //
        // Therefore, we fix this by creating a new range starting at 0 up to
        // the source length - 1.
        //
        // TODO: Create the proper range here.
        let src_len = self.source.len();
        let mut range = ranged.range();
        if range.start().to_usize() > src_len || range.end().to_usize() > src_len {
            range = TextRange::new(0.into(), self.source.text_len() - TextSize::from(1));
        }
        &self.source[range]
    }

    /// Parses elements enclosed within a delimiter pair, such as parentheses, brackets,
    /// or braces.
    fn parse_delimited(
        &mut self,
        allow_trailing_delim: bool,
        opening: TokenKind,
        delim: TokenKind,
        closing: TokenKind,
        func: impl FnMut(&mut Parser<'src>),
    ) {
        self.bump(opening);

        self.parse_separated(allow_trailing_delim, delim, [closing].as_slice(), func);

        self.expect_and_recover(closing, TokenSet::EMPTY);
    }

    /// Parses a sequence of elements separated by a delimiter. This function stops
    /// parsing upon encountering any of the tokens in `ending_set`, if it doesn't
    /// encounter the tokens in `ending_set` it stops parsing when seeing the `EOF`
    /// or `Newline` token.
    fn parse_separated(
        &mut self,
        allow_trailing_delim: bool,
        delim: TokenKind,
        ending_set: impl Into<TokenSet>,
        mut func: impl FnMut(&mut Parser<'src>),
    ) {
        let ending_set = NEWLINE_EOF_SET.union(ending_set.into());
        let mut progress = ParserProgress::default();

        while !self.at_ts(ending_set) {
            progress.assert_progressing(self);
            func(self);

            // exit the loop if a trailing `delim` is not allowed
            if !allow_trailing_delim && ending_set.contains(self.peek_nth(1)) {
                break;
            }

            if !self.eat(delim) {
                if self.at_expr() {
                    self.expect(delim);
                } else {
                    break;
                }
            }
        }
    }

    fn is_current_token_postfix(&self) -> bool {
        matches!(
            self.current_kind(),
            TokenKind::Lpar | TokenKind::Lsqb | TokenKind::Dot | TokenKind::Async | TokenKind::For
        )
    }

    fn handle_unexpected_indentation(&mut self, stmts: &mut Vec<Stmt>, error_msg: &str) {
        self.bump(TokenKind::Indent);

        self.add_error(
            ParseErrorType::OtherError(error_msg.to_string()),
            self.current_range(),
        );

        let mut progress = ParserProgress::default();

        while !self.at(TokenKind::Dedent) && !self.at(TokenKind::EndOfFile) {
            progress.assert_progressing(self);

            let stmt = self.parse_statement();
            stmts.push(stmt);
        }

        assert!(self.eat(TokenKind::Dedent));
    }
}

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
enum SequenceMatchPatternParentheses {
    Tuple,
    List,
}

impl SequenceMatchPatternParentheses {
    fn closing_kind(self) -> TokenKind {
        match self {
            SequenceMatchPatternParentheses::Tuple => TokenKind::Rpar,
            SequenceMatchPatternParentheses::List => TokenKind::Rsqb,
        }
    }

    const fn is_list(self) -> bool {
        matches!(self, SequenceMatchPatternParentheses::List)
    }
}

bitflags! {
    #[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
    struct ParserCtxFlags: u8 {
        const PARENTHESIZED_EXPR = 1 << 0;

        // NOTE: `ARGUMENTS` can be removed once the heuristic in `parse_with_items`
        // is improved.
        const ARGUMENTS = 1 << 1;
        const FOR_TARGET = 1 << 2;
    }
}

#[derive(PartialEq, Copy, Clone)]
enum FunctionKind {
    Lambda,
    FunctionDef,
}
