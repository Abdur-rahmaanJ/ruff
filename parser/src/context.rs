use crate::ast::{self, Expr, ExprContext, ExprKind};

pub(crate) fn set_context(expr: Expr, ctx: ExprContext) -> Expr {
    match expr.node {
        ExprKind::Name(ast::ExprName { id, .. }) => Expr {
            node: ast::ExprName { id, ctx }.into(),
            ..expr
        },
        ExprKind::Tuple(ast::ExprTuple { elts, .. }) => Expr {
            node: ast::ExprTuple {
                elts: elts
                    .into_iter()
                    .map(|elt| set_context(elt, ctx.clone()))
                    .collect(),
                ctx,
            }
            .into(),
            ..expr
        },
        ExprKind::List(ast::ExprList { elts, .. }) => Expr {
            node: ast::ExprList {
                elts: elts
                    .into_iter()
                    .map(|elt| set_context(elt, ctx.clone()))
                    .collect(),
                ctx,
            }
            .into(),
            ..expr
        },
        ExprKind::Attribute(ast::ExprAttribute { value, attr, .. }) => Expr {
            node: ast::ExprAttribute { value, attr, ctx }.into(),
            ..expr
        },
        ExprKind::Subscript(ast::ExprSubscript { value, slice, .. }) => Expr {
            node: ast::ExprSubscript { value, slice, ctx }.into(),
            ..expr
        },
        ExprKind::Starred(ast::ExprStarred { value, .. }) => Expr {
            node: ast::ExprStarred {
                value: Box::new(set_context(*value, ctx.clone())),
                ctx,
            }
            .into(),
            ..expr
        },
        _ => expr,
    }
}

#[cfg(test)]
mod tests {
    use crate::parser::parse_program;

    #[test]
    fn test_assign_name() {
        let source = "x = (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_tuple() {
        let source = "(x, y) = (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_list() {
        let source = "[x, y] = (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_attribute() {
        let source = "x.y = (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_subscript() {
        let source = "x[y] = (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_starred() {
        let source = "(x, *y) = (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_for() {
        let source = "for x in (1, 2, 3): pass";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_list_comp() {
        let source = "x = [y for y in (1, 2, 3)]";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_set_comp() {
        let source = "x = {y for y in (1, 2, 3)}";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_with() {
        let source = "with 1 as x: pass";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_assign_named_expr() {
        let source = "if x:= 1: pass";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_ann_assign_name() {
        let source = "x: int = 1";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_aug_assign_name() {
        let source = "x += 1";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_aug_assign_attribute() {
        let source = "x.y += (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_aug_assign_subscript() {
        let source = "x[y] += (1, 2, 3)";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_del_name() {
        let source = "del x";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_del_attribute() {
        let source = "del x.y";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }

    #[test]
    fn test_del_subscript() {
        let source = "del x[y]";
        let parse_ast = parse_program(source, "<test>").unwrap();
        insta::assert_debug_snapshot!(parse_ast);
    }
}
