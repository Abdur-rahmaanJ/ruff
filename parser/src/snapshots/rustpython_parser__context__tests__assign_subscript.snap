---
source: parser/src/context.rs
expression: parse_ast
---
[
    Attributed {
        range: 0..16,
        custom: (),
        node: Assign(
            StmtAssign {
                targets: [
                    Attributed {
                        range: 0..4,
                        custom: (),
                        node: Subscript(
                            ExprSubscript {
                                value: Attributed {
                                    range: 0..1,
                                    custom: (),
                                    node: Name(
                                        ExprName {
                                            id: "x",
                                            ctx: Load,
                                        },
                                    ),
                                },
                                slice: Attributed {
                                    range: 2..3,
                                    custom: (),
                                    node: Name(
                                        ExprName {
                                            id: "y",
                                            ctx: Load,
                                        },
                                    ),
                                },
                                ctx: Store,
                            },
                        ),
                    },
                ],
                value: Attributed {
                    range: 7..16,
                    custom: (),
                    node: Tuple(
                        ExprTuple {
                            elts: [
                                Attributed {
                                    range: 8..9,
                                    custom: (),
                                    node: Constant(
                                        ExprConstant {
                                            value: Int(
                                                1,
                                            ),
                                            kind: None,
                                        },
                                    ),
                                },
                                Attributed {
                                    range: 11..12,
                                    custom: (),
                                    node: Constant(
                                        ExprConstant {
                                            value: Int(
                                                2,
                                            ),
                                            kind: None,
                                        },
                                    ),
                                },
                                Attributed {
                                    range: 14..15,
                                    custom: (),
                                    node: Constant(
                                        ExprConstant {
                                            value: Int(
                                                3,
                                            ),
                                            kind: None,
                                        },
                                    ),
                                },
                            ],
                            ctx: Load,
                        },
                    ),
                },
                type_comment: None,
            },
        ),
    },
]
